Update Azure Governance Policy api-version to 2018-03-01

mentat9 opened the pull request:
This api spec combines api-version 2017-06-01-preview and 2016-12-01 into one, cleans up all validation issues, adds full examples, and clarifies/improves all descriptions. It should be fully back-compatible.

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

Commit:
Update swagger for api-version 2018-03-01
 - add many missing examples
 - improve/complete descriptions and summaries

Commit:
Fix a handful of validation failures
Add/correct some response codes

AutorestCI commented:
# Automation for azure-sdk-for-node

marstr reviewed (changes requested):
Our CI system is pretty unhappy with this PR. Before I review this in earnest, could you fix the invalid JSON that is likely messing with our tooling?
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/376107691#L699

AutorestCI commented:
# Automation for azure-libraries-for-java

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-python

mentat9 commented:
I'd love to. Thing is, that file exists and is completely valid JSON: https://github.com/mentat9/azure-rest-api-specs/blob/master/specification/resources/resource-manager/Microsoft.Authorization/stable/2018-03-01/examples/listBuiltInPolicyDefinitions.json

I haven't learned all the quirks of Travis, but it looks like it's detecting something different than what the message says. I don't see anything wrong with that file, or anything wrong in the nearby JSON. All of the local validation tools find that file and validate it just fine. Have you come across this kind of error before?

mentat9 commented:
Adding @pilor

Commit:
Fix back-compat for some error codes (leaving it inconsistent)
Change policyAssignment result for nonexistent delete by ID to match all other delete methods
Attempt to fix Travis FNF error

Commit:
Fix bracket bug in last commit

Commit:
Another attempt to fix Travis CI FNF error: simplifying the example file as much as possible.

Commit:
Another attempt to fix Travis FNF error (filename casing...)

Commit:
Putting back original example, modified during troubleshooting.

marstr commented:
> Autorest running against a local repo on my dev box resolves $ref as case INsensitive, but when it runs against a GitHub repo, it resolves that same $ref as case sensitive. This is a bad tool bug, IMO.

I understand that perspective, @mentat9! If you feel strongly enough, I'd recommend filing an issue over at [Azure/autorest](https://github.com/Azure/autorest).

mentat9 commented:
Done, thanks.

From: Martin Strobel [mailto:notifications@github.com]
Sent: Tuesday, May 8, 2018 3:48 PM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Chris Stackhouse <cstack@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] Update Azure Governance Policy api-version to 2018-03-01 (#3032)


Autorest running against a local repo on my dev box resolves $ref as case INsensitive, but when it runs against a GitHub repo, it resolves that same $ref as case sensitive. This is a bad tool bug, IMO.

I understand that perspective, @mentat9<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmentat9&data=02%7C01%7Ccstack%40microsoft.com%7Cec6a44ade7a34bfd001a08d5b535c879%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636614165269332737&sdata=7YUF6c8zASjkEP4aD%2FzFNEpktWRpzPl41aDMuxFu%2BZ0%3D&reserved=0>! If you feel strongly enough, I'd recommend filing an issue over at Azure/autorest<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fautorest&data=02%7C01%7Ccstack%40microsoft.com%7Cec6a44ade7a34bfd001a08d5b535c879%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636614165269332737&sdata=mRP8CLCVDzz9igt8lm3t098F4FGHCJ0e58Lc%2Blkeii8%3D&reserved=0>.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-rest-api-specs%2Fpull%2F3032%23issuecomment-387566476&data=02%7C01%7Ccstack%40microsoft.com%7Cec6a44ade7a34bfd001a08d5b535c879%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636614165269488986&sdata=kUoSVfyvIfIOLlFy%2Bmp3ce%2Fe4g78aEadf9KEqoqJR%2Bc%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAQDfR4o-ihc2ROBBdb6EsmkuesJqdeyGks5twiCugaJpZM4T1tXL&data=02%7C01%7Ccstack%40microsoft.com%7Cec6a44ade7a34bfd001a08d5b535c879%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636614165269488986&sdata=ahXgNCAdB01sezGgJA5nE8AELE%2FCvQenV5DX5oxHND8%3D&reserved=0>.

Commit:
Apply PR feedback (fix errors in a couple examples)

ravbhatnagar commented:
@marstr @mentat9 - Its hard to review the changes in this api-version. The first commit didnt include the original version.

ravbhatnagar commented:
Ok, so I compared the policyAssignment.json, policyDefinition.json with the previos version using a diff tool. There were no major changes. I also took a look at policySetDefinition.json which is a new file. APIs for this are already live on service side. So signing off from ARM side.

marstr commented:
@mentat9, are these changes deployed from your side, and ready for me to merge?

mentat9 commented:
Yes, we are good to go. Please merge at your earliest convenience.

Commit:
Correct/improve wording in description properties.

mentat9 commented:
@marstr, We are good to go from a service perspective, and my description updates are in. Please merge at your earliest convenience after checks succeed.

mentat9 commented:
@marstr do you happen to know whom I should follow this failure up with? I need to generate the Node.js SDK, but the docs give an email alias that no longer exists for the Node.js SDK team.

> # Automation for azure-sdk-for-node
> Encountered a Subprocess error: (azure-sdk-for-node)
> 
> Command: ['/usr/local/bin/autorest', '/tmp/tmp6rn5la_d/rest/specification/resources/resource-manager/readme.md', '--license-header=MICROSOFT_MIT_NO_VERSION', '--node-sdks-folder=/tmp/tmp6rn5la_d/sdk', '--nodejs', '--use=@microsoft.azure/autorest.nodejs@^2.1.43']
> Finished with return code 1
> and output:
> AutoRest code generation utility [version: 2.0.4262; node: v7.10.1]
> (C) 2018 Microsoft Corporation.
> https://aka.ms/autorest
>    Loading AutoRest core      '/root/.autorest/@microsoft.azure_autorest-core@2.0.4275/node_modules/@microsoft.azure/autorest-core/dist' (2.0.4275)
>    Loading AutoRest extension '@microsoft.azure/autorest.nodejs' (^2.1.43->2.1.59)
>    Loading AutoRest extension '@microsoft.azure/autorest.modeler' (2.3.44->2.3.44)
> [Exception] No input files provided.
> 
> Use --help to get help information.

RikkiGibson commented:
@AutorestCI rebuild azure-sdk-for-node

AutorestCI commented:
<details><summary>Encountered a Subprocess error</summary><p>

Command: ['/usr/local/bin/autorest', '/tmp/tmpda054lkj/rest/specification/resources/resource-manager/readme.md', '--license-header=MICROSOFT_MIT_NO_VERSION', '--node-sdks-folder=/tmp/tmpda054lkj/sdk', '--nodejs', '--use=@microsoft.azure/autorest.nodejs@^2.1.43']
Finished with return code 1
and output:
```shell
AutoRest code generation utility [version: 2.0.4262; node: v7.10.1]
(C) 2018 Microsoft Corporation.
https://aka.ms/autorest
   Loading AutoRest core      '/root/.autorest/@microsoft.azure_autorest-core@2.0.4278/node_modules/@microsoft.azure/autorest-core/dist' (2.0.4278)
   Loading AutoRest extension '@microsoft.azure/autorest.nodejs' (^2.1.43->2.1.59)
   Loading AutoRest extension '@microsoft.azure/autorest.modeler' (2.3.44->2.3.44)
[Exception] No input files provided.

Use --help to get help information.
```

</p></details>

RikkiGibson commented:
@AutorestCI rebuild azure-sdk-for-node

AutorestCI commented:
<details><summary>Encountered a Subprocess error</summary><p>

Command: ['/usr/local/bin/autorest', '/tmp/tmpnpmjnkn8/rest/specification/resources/resource-manager/readme.md', '--license-header=MICROSOFT_MIT_NO_VERSION', '--node-sdks-folder=/tmp/tmpnpmjnkn8/sdk', '--nodejs', '--use=@microsoft.azure/autorest.nodejs@^2.1.43']
Finished with return code 1
and output:
```shell
AutoRest code generation utility [version: 2.0.4280; node: v8.11.3]
(C) 2018 Microsoft Corporation.
https://aka.ms/autorest
   Loading AutoRest core      '/root/.autorest/@microsoft.azure_autorest-core@2.0.4280/node_modules/@microsoft.azure/autorest-core/dist' (2.0.4280)
   Loading AutoRest extension '@microsoft.azure/autorest.nodejs' (^2.1.43->2.1.59)
   Loading AutoRest extension '@microsoft.azure/autorest.modeler' (2.3.44->2.3.44)
[Exception] No input files provided.

Use --help to get help information.
```

</p></details>

RikkiGibson commented:
Really not sure what the problem is-- when I copy those args, modify to my local versions of the SDK and rest-api-specs and run locally, it's fine. I'm opening a PR for you @mentat9.
